ccm_web/server/src/canvas/canvas.service.ts

```diff
@@ -27,34 +27,35 @@ interface IncomingRateLimitedCanvasHttpHeaders extends IncomingHttpHeaders {
 
 const requestorOptions: GotOptions = {
   retry: {
-    limit: 2,
+    limit: 3,
     methods: ['POST', 'GET', 'PUT', 'DELETE'],
-    statusCodes: got.defaults.options.retry.statusCodes.concat([403])
+    statusCodes: got.defaults.options.retry.statusCodes.concat([403]),
+    calculateDelay: ({ attemptCount, retryOptions, error, computedValue }) => {
+      const headers = error.response?.headers as IncomingRateLimitedCanvasHttpHeaders
+      const delay: number = computedValue === 0 ? 0 : 5000
```

Comment:
I did some more testing with your code, and I encountered what I think you are seeing a few times. Basically you get receive `error` that isn't an `HTTPError` and doesn't have a status code but doesn't have a `code` either (like `ECONNRESET`). I'm not sure our configuration is set up to retry when this occurs.

There's another issue I saw that in these cases, you're still trying to access these custom headers, so there's an error message to the effect of the "property x-rate-limit-remaining doesn't exist on object undefined". This may be why you're not seeing `calculateDelay` in these cases, but I'm not sure about that.

Anyways, I'm going to do some more testing on Monday with something like this:
```
  retry: {
    limit: 3,
    methods: ['POST', 'GET', 'PUT', 'DELETE'],
    statusCodes: got.defaults.options.retry.statusCodes.concat([403]),
    calculateDelay: ({ attemptCount, retryOptions, error, computedValue }) => {
      const headers = error.response?.headers
      const isStrangeError = !(error instanceof HTTPError) && error.code === undefined
      logger.debug(`isStrangeError: ${String(isStrangeError)}`)
      const delay = computedValue === 0 && !isStrangeError ? 0 : 5000
      logger.debug(
        `calculateDelay [${String(attemptCount)}] — ` +
        `"delay": "${String(delay)}"; ` +
        `"retryOptions": "${JSON.stringify(retryOptions)}";` +
        `error.code: ${String(error.code)}`
      )
      if (headers !== undefined) {
        logger.debug(
          `"x-rate-limit-remaining": "${String(headers['x-rate-limit-remaining'])}"; ` +
          `"x-request-cost": "${String(headers['x-request-cost'])}"`
        )
      }
      return delay
    }
  },
```

Reply:
@lsloan, before you try proactive request limiting, can you see whether this helps at all for you?

Reply:
I'll give that a shot.

Interesting about the log messages with Canvas API header values.  I thought TS/JS would return a string like "undefined" for those.

Reply:
Thanks!  That's a valuable TS lesson.